packages/astro/src/webcontainer-files/filesmap.ts

```diff
@@ -0,0 +1,218 @@
+
+      if (!dependOn) {
+        dependOn = new FilesMap(dependOnPath);
+        graph.set(dependOnPath, dependOn);
+        dependOn.update(graph, logger);
+      }
+
+      this.extend(dependOn);
+    } catch (error: any) {
+      logger.warn(`Failed to parse '${configPath}', content won't be included.\nError: ${error}`);
+    }
+  }
+
+  extend(other: FilesMap) {
+    this.unlink();
+    other._dependents.add(this);
+    this._extend = other;
+  }
+
+  unlink() {
+    this._extend?._dependents.delete(this);
+    this._extend = null;
+  }
+
+  *allDependents(): Generator<FilesMap> {
+    for (const dependency of this._dependents) {
+      yield dependency;
+
+      yield* dependency.allDependents();
+    }
+  }
+
+  async toFiles(logger: AstroIntegrationLogger): Promise<Files> {
+    const filePathsObj = await this._toFilePathsCheckCycles({
+      logger,
+      visitedNodes: [],
+    });
+
+    delete filePathsObj[EXTEND_CONFIG_FILEPATH];
+
+    const filePaths = Object.entries(filePathsObj);
+
+    filePaths.sort();
+
+    const files: Files = {};
+
+    for (const [webcontainerPath, filePath] of filePaths) {
+      const buffer = fs.readFileSync(filePath);
+
+      try {
+        const stringContent = new TextDecoder('utf-8', { fatal: true }).decode(buffer);
+
+        files[webcontainerPath] = stringContent;
+      } catch {
+        files[webcontainerPath] = { base64: buffer.toString('base64') };
+      }
+    }
+
+    return files;
+  }
+
+  private async _toFilePathsCheckCycles(context: CheckCycleContext): Promise<FilePaths> {
+    const seenIndex = context.visitedNodes.indexOf(this.path);
+
+    if (seenIndex !== -1) {
+      context.logger.warn(
+        [
+          'Cycle detected:',
+          ...context.visitedNodes.map((filePath, index) => {
+            return `  * '${filePath}' ${index === seenIndex ? '<-- Cycle points back to that file' : ''}`;
+          }),
+          'Content will be ignored after that point.',
+        ].join('\n'),
+      );
+
+      return {};
+    }
+
+    context.visitedNodes.push(this.path);
+
+    const filePaths: FilePaths = this._extend ? await this._extend._toFilePathsCheckCycles(context) : {};
+
+    await getAllFiles(this.path, filePaths);
+
+    return filePaths;
+  }
+}
+
+interface CheckCycleContext {
+  logger: AstroIntegrationLogger;
+  visitedNodes: string[];
+}
+
+type FilePaths = { [webcontainerPath: string]: /* filePath: */ string };
+
+async function getAllFiles(dir: string, result: FilePaths): Promise<void> {
+  const filePaths = await glob(`${glob.convertPathToPattern(dir)}/**/*`, {
+    onlyFiles: true,
+    dot: true,
+    ignore: IGNORED_FILES,
+  });
+
+  for (const filePath of filePaths) {
+    result[webcontainerPath(dir, filePath)] = filePath;
+  }
+}
+
+function webcontainerPath(dir: string, filePath: string) {
+  const result = `/${path.relative(dir, filePath)}`;
+
+  // mostly for Windows
+  if (path.sep !== '/') {
+    return result.replaceAll('\\', '/');
+  }
```

Comment:
I guess `path.sep` can only be `/` or `\` but following would handle all possible edge cases:

```suggestion
  // normalize path separators
  if (path.sep !== '/') {
    return result.replaceAll(path.sep, '/');
  }
```

Reply:
Good point!